---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  cache = TRUE
)
```

# mousePreopticR

<!-- badges: start -->
<!-- badges: end -->

The goal of `mousePreopticR` is to enable more convenient sharing/loading of bioinformatics gene expression datasets for analysis in R, in particular mouse hypothalamic preoptic region cell data from experiments by Moffit et al. The data can be found at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE113576) and [DRYAD](https://datadryad.org/stash/dataset/doi:10.5061/dryad.8t8s248).

## Installation

You can install the development version of `mousePreopticR` from [GitHub](https://github.com/) with:

```{r github_install, message=FALSE, warning=FALSE, eval=FALSE}
# install.packages("devtools")
devtools::install_github("wjonasreger/mousePreopticR")
```

Installing the package using `devtools::install_github` might have issues due to memory size of package data or rate limits with GitHub. Alternatively, you can download `mousePreopticR` [here](https://github.com/wjonasreger/mousePreopticR/archive/refs/heads/master.zip). 

You will need to unzip the package folder, remove _"-master"_ from the package folder name, and relocate it to your desired directory, then install it with the following code using _your own path to the package_.

```{r local_install, message=FALSE, eval=FALSE}
install.packages("/Users/wjonasreger/Documents/mousePreopticR", repos = NULL, type="source")
```

## Usage

This is a basic example which shows you how to setup the package datasets for analysis:

***NOTE:** The package datasets are split components of sample datasets due to GitHub size limits. So, the documented datasets (i.e., `mpr_barcodes`, `mpr_genes`, `mpr_matrix`, `mpr_merfish`) need to be composed once after each install/update of the package. Then package datasets can be loaded anytime after the initial composition step. You may need to restart your R session after composition is completed.*

### Import packages

```{r packages, message=FALSE}
## import packages
library(mousePreopticR)
library(Seurat)
library(monocle3)
library(SpatialExperiment)
library(scran)
library(ggspavis)
library(nnSVG)
library(dplyr)
```


### Load Data

```{r data, message=FALSE}
## compose datasets
#     - only need to run this once after installation. 
#     - may need to restart R session after composition.
composeData()


## load package datasets
#     - mpr_matrix, mpr_genes, mpr_barcodes
#     - mpr_merfish
loadData()


## load local data as expression matrix
#     - requires "matrix.mtx", "genes.tsv", and "barcodes.tsv" in data directory
# mpr_matrix = readMPR(data_dir = "mpr_experiment", cell_column = 1, gene_column = 1)


## viewing data
# to learn more about a dataset, use ?data_name (e.g., ?mpr_matrix)
?mpr_matrix
```


### Gene Expression Matrix Data Objects

```{r expression_matrix_objects, message=FALSE}
## Seurat object with mpr_matrix (Seurat)
mpr_seurat = CreateSeuratObject(counts = mpr_matrix)
mpr_seurat


## CellDataSet object with mpr_matrix, mpr_barcodes, and mpr_genes (monocle3)
cds = new_cell_data_set(expression_data = mpr_matrix,
                        cell_metadata = mpr_barcodes,
                        gene_metadata = mpr_genes)
cds
```


```{r spatial_experiment_objects, message=FALSE}
## SpatialExperiment object with mpr_merfish (SpatialExperiment)
# view documentation of mpr_merfish data with spatial features
?mpr_merfish

# preprocess data
# define names for columns, rows, and meta data for intended matrix
names = list(
  col = colnames(mpr_merfish),
  row = mpr_merfish$Cell_ID,
  meta = c("Cell_ID", "Centroid_X", "Centroid_Y")
)

# define subset conditions
subs = list(
  col = setdiff(names[["col"]][10:170], "Fos"), # ignore Fos column due to NAs
  row = names[["row"]]
)

# convert merfish tabular data to matrix object with associated metadata
mpr_mf_mat = matrixify(data = mpr_merfish, names, subs, transpose = TRUE)

# define features for SpatialExperiment object
# expression matrix
counts = mpr_mf_mat[["matrix"]]
# row data
row_data = data.frame(gene_short_name = counts@Dimnames[[1]])
rownames(row_data) = row_data$gene_short_name
# column data
col_data = data.frame(barcode = counts@Dimnames[[2]])
rownames(col_data) = col_data$barcode
# spatial coordinates
spatial_coords = as.matrix(apply(as.data.frame(mpr_mf_mat[["metadata"]][3:4]), 2, as.numeric))

# create SpatialExperiment
mpr_spe = SpatialExperiment(
  assays = list(counts = counts),
  rowData = row_data,
  colData = col_data,
  spatialCoords = spatial_coords
)
mpr_spe
```

## Example Analysis of Spatial Experiment Data

```{r spe_prep}
# preprocessing
col_sums = colSums(counts(mpr_spe))
mpr_spe = mpr_spe[, col_sums > 0]
mpr_spe = logNormCounts(mpr_spe)
```

```{r spe_cluster, warning=FALSE, cache = TRUE}
# graph-based clustering
set.seed(22)
k = 50
g = buildSNNGraph(mpr_spe, k = k, use.dimred = NULL)
g_walk = igraph::cluster_walktrap(g)
clus = g_walk$membership
table(clus)
```

```{r spe_plot_clusters, cache = TRUE}
# store cluster labels in column 'label' in colData
colLabels(mpr_spe) = factor(clus)

# salmon sushi color palette
paletteFn = colorRampPalette(c("#E7F2F8", "#74BDCB", "#FFA384", "#EFE7BC"))

# plot clusters in spatial x-y coordinates
plotSpots(mpr_spe, annotate = "label", in_tissue = NULL,
          palette = paletteFn(length(unique(clus))))
```

```{r spe_diff_exp}
# convert from spe to so
colnames(mpr_spe) = 1:dim(mpr_spe)[2]
spe_seurat = CreateSeuratObject(counts = counts(mpr_spe))
Idents(spe_seurat) = mpr_spe$label

# find all differentially expressed genes across clusters
spe_markers = FindAllMarkers(spe_seurat, verbose = FALSE)

# differentially expressed genes
diff_exp = spe_markers %>%
  group_by(cluster) %>%
  filter(avg_log2FC >= 1.5) %>%
  filter(p_val_adj <= 0.05) %>%
  arrange(desc(avg_log2FC))
diff_exp
```


```{r spe_plot_top_gene}
# plot top gene
mpr_spe[["gene_x"]] = log(counts(mpr_spe)["Selplg",])
plotSpots(mpr_spe, annotate = "gene_x", in_tissue = NULL,
          palette = "libd_layer_colors")
```



Ok, that's all! You can now play _[scientist](https://youtu.be/RB-RcX5DS5A)_ and do fancy bioinformatics analysis with the `mousePreopticR` datasets, or even use it's convenience functions to manage other datasets as you wish... but that's probably _a lot_ of work. There are more exquisite ways to spend your time such as getting coffee with your friend ðŸ˜„â˜• â˜•ðŸ˜„

...You're still here? Ok [watch this](https://youtu.be/xvFZjo5PgG0) for a cackle ;)
